File: src/main/java/com/zihao/Shenzhen.java
package com.zihao;


import java.util.List;
import java.util.Objects;
import java.util.concurrent.TimeUnit;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;


/**
 * @author szq
 * Created on 2023-08-01
 * 使用Java+Selenium 自动框架实现
 */
public class Shenzhen {

    public static void main(String[] args) throws Exception {


        // 1、设置chrome浏览器驱动
        System.setProperty("webdriver.chrome.driver", "/Users/shenzhenqiang/Downloads/chromedriver_mac_arm64/chromedriver");


        // 2、设置ChromeOptions
        ChromeOptions chromeOptions = initChromeOptions();
        chromeOptions.addArguments(
                "user-agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36");
        //        chromeOptions.addArguments("--headless");//无头浏览
        //        chromeOptions.addArguments("blink-settings=imagesEnabled=false");//禁用图片
        // 3、初始化一个浏览器实例
        WebDriver webDriver = initWebDriver(chromeOptions);
        //        chromeOptions.addArguments("--user-agent=Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/58.0.3029.110 Safari/537.36,"
        //                + "--cookie=");

        //         4、加载网页程序（某度）
        webDriver.get("https://www.shenzhenair.com/szair_B2C");
        //*[@id="orgCity"]

        WebElement wait = webDriver.findElement(By.className("layui-layer-btn1"));
        wait.click();
        WebElement orgCity = webDriver.findElement(By.id("orgCity"));
        WebElement dstCity = webDriver.findElement(By.id("dstCity"));
        WebElement orgDate = webDriver.findElement(By.id("orgDate"));

        String orgCit = "北京首都";
        String dstCit = "深圳";

        orgCity.sendKeys(orgCit);
        orgDate.clear();
        orgDate.sendKeys("2023-08-16");

        dstCity.sendKeys(dstCit);

        WebElement submitButton = webDriver.findElement(By.id("search-flight"));
        submitButton.click();

        System.out.println("深圳航空");

        List<WebElement> flightList = webDriver.findElements(By.cssSelector("#flightInfoListDC > table > tbody > tr"));
        System.out.println(flightList.size());
        flightList.forEach(flight -> {
            String judge = flight.getAttribute("class");
            //            System.out.println(flight.getAttribute("class"));
            String id = flight.getAttribute("id");
            if (!judge.equals("flightTr")) {
                return;
            }

            //             航班号
            WebElement flightNumber1 = flight.findElement(By.xpath("//*[@id=\"" + id + "\"]" + "/td[1]/div[1]"));
            System.out.println("航班号:" + flightNumber1.getText());

            WebElement departTime = flight.findElement(By.xpath("//*[@id=\"" + id + "\"]" + "/td[1]/div[2]"));
            System.out.println("起飞落地时间: " + departTime.getText());

            // 起飞机场
            WebElement departAirport =
                    //                    #flightInfogo1 > td.flightInfoForm > div.F16.detailInfo.tipso_style
                    //*[@id="flightInfogo1"]/td[1]/div[3]
                    flight.findElement(By.xpath("//*[@id=\"" + id + "\"]" + "/td[1]/div[3]"));
            System.out.println("机场起点: " + departAirport.getAttribute("row1-from"));
            // 如果有中转
            if(Objects.nonNull(departAirport.getAttribute("row2-info"))) {
                System.out.println(departAirport.getAttribute("row2-info"));
                System.out.println("机场终点：" + departAirport.getAttribute("row3-to"));
            } else {  //否则没有中转
                System.out.println("机场终点： " + departAirport.getAttribute("row1-to"));
            }

            try {
                WebElement headPrice = flight.findElement(By.cssSelector("td:nth-child(2) > div.F22.notHover"));
                System.out.println("头等舱价格:" + headPrice.getText());
                WebElement headSeatNumbers = flight.findElement(By.cssSelector("td:nth-child(2) > div.F16.notHover"));
                System.out.println("头等舱座位:" + headSeatNumbers.getText());
            } catch (Exception e) {
                System.out.println("————————————没有头等舱——————————————");
            }
            try {
                WebElement superPrice = flight.findElement(By.cssSelector("td:nth-child(3) > div.F22.notHover"));
                System.out.println("超值公务舱价格:" + superPrice.getText());
                WebElement superSeatNumbers = flight.findElement(By.cssSelector("td:nth-child(3) > div.F16.notHover"));
                System.out.println("超值公务舱座位:" + superSeatNumbers.getText());
            } catch (Exception e) {
                System.out.println("————————————没有超值公务舱——————————————");
            }
            try {
                WebElement comfortablePrice = flight.findElement(By.cssSelector("td:nth-child(4) > div.F22.notHover"));
                System.out.println("舒适经济舱价格:" + comfortablePrice.getText());
                WebElement comfortableSeatNumbers = flight.findElement(By.cssSelector("td:nth-child(4) > div.F16.notHover"));
                System.out.println("舒适经济舱座位:" + comfortableSeatNumbers.getText());
            } catch (Exception e) {
                System.out.println("————————————没有舒适经济舱——————————————");
            }
            try {
                WebElement cheapPrice = flight.findElement(By.cssSelector("td:nth-child(5) > div.F22.notHover"));
                System.out.println("经济舱价格:" + cheapPrice.getText());

                WebElement cheapSeatNumbers = flight.findElement(By.cssSelector("td:nth-child(5) > div.F16.notHover"));
                System.out.println("经济舱座位:" + cheapSeatNumbers.getText());
            } catch (Exception e) {
                System.out.println("————————————没有经济舱——————————————");
            }
            System.out.println("-------------------------------------------------------------");
            //*[@id="flightInfogo0"]/td[3]/div[2]
            //            // 到达时间
            //            WebElement arriveTime = flArrive.findElement(By.className("hours")).findElement(By.cssSelector("span"));
            //            System.out.println(arriveTime);
            //            // 到达机场
            //            WebElement arriveAirport = flArrive.findElement(By.className("airport"));
            //            System.out.println(arriveAirport);

            //            WebElement flCenter = flight.findElement(By.className("fl-center"));
            //            // 飞行时间
            //            WebElement durationTime = flCenter.findElement(By.className("durationTime"));
            //            System.out.println(durationTime);
            //
            //            WebElement price = flight.findElement(By.className("price"));
            //            // 价格
            //            WebElement priceNum = price.findElement(By.className("num"));
            //            System.out.println(priceNum);

            //
            //            Flight flightObject = Flight.builder()
            ////                    .airCom(flightCompany.getText())
            //                    .flightNumber(flightNumber.getText())
            //                    .departTime(departTime.getText())
            //                    .departAirport(departAirport.getText())
            //                    .arriveTime(arriveTime.getText())
            //                    .arriveAirport(arriveAirport.getText())
            //                    .durationTime(durationTime.getText())
            //                    .price(priceNum.getText()).build();

            //            System.out.println(JsonUtil.toJSON(flightObject));

        });
        webDriver.close();


    }

    private static ChromeOptions initChromeOptions() {
        ChromeOptions options = new ChromeOptions();
            /*// 启动时自动最大化窗口
            options.addArguments("--start-maximized");
            // 禁用阻止弹出窗口
            options.addArguments("--disable-popup-blocking");
            // 启动无沙盒模式运行
            options.addArguments("no-sandbox");
            // 禁用扩展
            options.addArguments("disable-extensions");
            // 默认浏览器检查
            options.addArguments("no-default-browser-check");
            // 设置chrome 无头模式
            options.setHeadless(Boolean.TRUE);
            //不用打开图形界面。
            options.addArguments("--headless");*/
        options.setBinary("/Applications/Google Chrome.app/Contents/MacOS/Google Chrome");
        return options;
    }

    private static WebDriver initWebDriver(ChromeOptions options) {
        WebDriver webDriver = new ChromeDriver(options);
        webDriver.manage().window().maximize();
        webDriver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
        return webDriver;
    }
}

File: src/main/java/com/zihao/python/shen_zhen.py
# coding=utf-8
# @Time    : 2023/8/5--10:32
# @Author  : child_lyj
# @File    : main
# @Software: PyCharm
# @Project : Spider
# @Description:
import time

from selenium import webdriver
from selenium.webdriver.common.by import By


options = webdriver.ChromeOptions()
user_agent='user-agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36'
options.add_argument(f'user-agent={user_agent}')
driver = webdriver.Chrome(options=options)
driver.get("https://www.shenzhenair.com/szair_B2C/")

time.sleep(3)
element = driver.find_element(By.CLASS_NAME, "layui-layer-btn1")  # 我知道了按钮
# print(element)
element.click()  # 点击关闭弹窗
element = driver.find_element(By.ID, "orgCity")  # 始发地输入框
# print(element)
element.send_keys("北京首都")
element = driver.find_element(By.ID, "dstCity")  # 终到地输入框
# print(element)
element.send_keys("深圳")
element = driver.find_element(By.ID, "orgDate")  # 触发日期输入框
# print(element)
element.clear()
element.send_keys("2023-08-14")
element = driver.find_element(By.CLASS_NAME, "ydxc")  # 预订行程按钮，用于隐藏日期选择框，否则会挡住查询按钮
element.click()
element = driver.find_element(By.ID, "search-flight")  # 立即预订查询按钮
# print(element)
element.click()
elements = driver.find_elements(By.CLASS_NAME, "flightTr")
# print(f"所有的信息:\n{elements}")

# 循环拿到每一趟航班的信息
for one_info in elements:
    # print(one_info)
    tds = one_info.find_elements(By.XPATH, "./td")  # 一趟航班所有的信息都放在6个<td>标签里面
    # print(tds)
    # print(len(tds))
    print(f"航班号: {tds[0].find_element(By.CLASS_NAME, 'F20').text}")  # 航班号
    print(f"航班时间: {tds[0].find_element(By.CLASS_NAME, 'F22').text}")  # 航班时间

    # 是否有中转
    is_transfer = False
    try:
        transfer_city = tds[0].find_element(By.CLASS_NAME, 'F20').find_element(By.CLASS_NAME, 'tipsoTing').get_attribute('routename') # 存在中转城市，则不会出现异常
        print(f"起点机场信息: {tds[0].find_element(By.CLASS_NAME, 'F16').get_attribute('row1-from')}")  # 机场信息
        print(f"中转机场信息: {tds[0].find_element(By.CLASS_NAME, 'F16').get_attribute('row1-to')}")  # 机场信息
        print(f"终点机场信息: {tds[0].find_element(By.CLASS_NAME, 'F16').get_attribute('row3-to')}")  # 机场信息
        is_transfer = True
    except:
        print(f"起点机场信息: {tds[0].find_element(By.CLASS_NAME, 'F16').get_attribute('row1-from')}")  # 机场信息
        print(f"中转机场信息: 无中转")  # 机场信息
        print(f"终点机场信息: {tds[0].find_element(By.CLASS_NAME, 'F16').get_attribute('row1-to')}")  # 机场信息
    try:
        print(f"头等/公务舱: {tds[1].find_element(By.CLASS_NAME, 'F22').text}, {tds[1].find_element(By.CLASS_NAME, 'F16').text}")  # 价格及座位信息
    except:
        print(f"头等/公务舱: 无该票务信息")
    try:
        print(f"超值公务舱: {tds[2].find_element(By.CLASS_NAME, 'F22').text}, {tds[2].find_element(By.CLASS_NAME, 'F16').text}")  # 价格及座位信息
    except:
        print(f"超值公务舱: 无该票务信息")
    try:
        print(f"舒适经济舱: {tds[3].find_element(By.CLASS_NAME, 'F22').text}, {tds[3].find_element(By.CLASS_NAME, 'F16').text}")  # 价格及座位信息
    except:
        print(f"舒适经济舱: 无该票务信息")
    try:
        print(f"经济舱: {tds[4].find_element(By.CLASS_NAME, 'F22').text}, {tds[4].find_element(By.CLASS_NAME, 'F16').text}")  # 价格及座位信息
    except:
        print(f"经济舱: 无该票务信息")

    print("-" * 80)  # 每一趟航班之间个分隔符
driver
